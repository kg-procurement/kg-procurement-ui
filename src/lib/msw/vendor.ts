import { http, HttpResponse } from 'msw'

import { API_BASE_URL } from '@/env.ts'
import { GetVendorByIdResponse, GetVendorsResponse, UpdateVendorResponse } from '@/lib/redux/features/vendor/validation.ts'

export const vendorHandlers = [
  http.get(`${API_BASE_URL}/vendor`, () => {
    return HttpResponse.json({
      vendors: [
        {
          id: '2502',
          name: 'Adriza Nikmah, PT',
          description: 'Headset dll',
          bp_id: '2502',
          bp_name: 'Adriza Nikmah, PT',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2507',
          name: 'Toko Amazon',
          description: 'Menjual segala macam Kebutuhan Dapur/ Home App...',
          bp_id: '2507',
          bp_name: 'Amazon',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2022-11-24T10:12:25Z',
          modified_by: '165',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2508',
          name: 'Aneka Warna',
          description: 'Vendor Sablon dan Garment',
          bp_id: '2508',
          bp_name: 'Aneka Warna',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2509',
          name: 'Angkasa Buana Niaga Cipta, PT',
          description: 'Headset Plantronics',
          bp_id: '2509',
          bp_name: 'Angkasa Buana Niaga Cipta, PT',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2511',
          name: 'Apartement Permata Senayan',
          description: '0',
          bp_id: '2511',
          bp_name: 'Apartement Permata Senayan',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2512',
          name: 'Arta Muara',
          description: 'Mesin Barcode',
          bp_id: '2512',
          bp_name: 'Arta Muara',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2513',
          name: 'Asaba Computer Centre',
          description: 'Miscrosoft Gold Partner',
          bp_id: '2513',
          bp_name: 'Asaba Computer Centre',
          rating: 1,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: '102722',
          modified_date: '2022-12-05T11:47:24Z',
          modified_by: '7',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2514',
          name: 'Astra Graphia TBK, PT',
          description: 'Distributor Xerox',
          bp_id: '2514',
          bp_name: 'Astra Graphia TBK, PT',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2515',
          name: 'Asus Service Center',
          description: 'Service Center dan Sales Spare Spart Asus',
          bp_id: '2515',
          bp_name: 'Asus Service Center',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
        {
          id: '2516',
          name: 'Atrindo Asia Global, PT',
          description: 'Penyemprotan Disinfectant',
          bp_id: '2516',
          bp_name: 'Atrindo Asia Global, PT',
          rating: 0,
          area_group_id: '1',
          area_group_name: 'Indonesia',
          sap_code: 'None',
          modified_date: '2020-10-29T13:39:12Z',
          modified_by: '0',
          dt: '2024-08-28T00:00:00Z',
        },
      ],
      metadata: {
        total_page: 10,
        current_page: 1,
        total_entries: 100,
      },
    } satisfies GetVendorsResponse)
  }),
  http.get(`${API_BASE_URL}/vendor/:id`, (req) => {
    const { id } = req.params
    if (id === '-1') {
      return HttpResponse.json(null, { status: 200 })
    }

    return HttpResponse.json(
      {
        id: '2502',
        name: 'Adriza Nikmah, PT',
        description: 'Headset dll',
        bp_id: '2502',
        bp_name: 'Adriza Nikmah, PT',
        rating: 0,
        area_group_id: '1',
        area_group_name: 'Indonesia',
        sap_code: 'None',
        modified_date: '2020-10-29T13:39:12Z',
        modified_by: '0',
        dt: '2024-08-28T00:00:00Z',
      } satisfies GetVendorByIdResponse,
      { status: 200 },
    )
  }),
  http.put(`${API_BASE_URL}/vendor/:id`, ({ params }) => {
    const { id } = params
    return HttpResponse.json({
      id: Array.isArray(id) ? id[0] : id,
      name: 'Adriza Nikmah, PT',
      description: 'Headset dll',
      bp_id: '2502',
      bp_name: 'Adriza Nikmah, PT',
      rating: 0,
      area_group_id: '1',
      area_group_name: 'Indonesia',
      sap_code: 'None',
      modified_date: '2020-10-29T13:39:12Z',
      modified_by: '0',
      dt: '2024-08-28T00:00:00Z',
    } satisfies UpdateVendorResponse)
  }),
]
