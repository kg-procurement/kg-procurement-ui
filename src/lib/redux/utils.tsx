import { SerializedError } from '@reduxjs/toolkit'
import { FetchBaseQueryError } from '@reduxjs/toolkit/query'

export type ZodParseError = Extract<
  FetchBaseQueryError,
  {
    status: 'PARSING_ERROR'
    originalStatus: number
    data: string
    error: string
  }
>

export function isZodParseError(
  err: SerializedError | FetchBaseQueryError | undefined,
): err is ZodParseError {
  if (!err) return false
  return (
    'data' in err &&
    'error' in err &&
    'originalStatus' in err &&
    'status' in err
  )
}

export function extractZodError(err: ZodParseError): React.ReactNode {
  const errorList = err.error.split('\n').slice(0, -1)
  return (
    <div className="flex flex-col">
      <div className="font-semibold">
        Response does not match validator schema:
      </div>
      <ul className="list-inside list-disc">
        {errorList.map((e, i) => (
          <li key={i}>{e}</li>
        ))}
      </ul>
    </div>
  )
}

export function isErrorWithMessage(
  err: FetchBaseQueryError,
): err is { status: number, data: { error: string } } {
  if (!err.data) return false
  return (
    typeof err.data === 'object' &&
    'error' in err.data &&
    typeof err.data.error === 'string'
  )
}
