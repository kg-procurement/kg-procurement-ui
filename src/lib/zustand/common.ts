import { create } from 'zustand'

export interface CommonStore {
  showLoadingOverlay: boolean
  setShowLoadingOverlay: (newValue: boolean) => void
}

export const useCommonStore = create<CommonStore>(set => ({
  showLoadingOverlay: false,
  setShowLoadingOverlay: newValue =>
    set(state => ({ ...state, showLoadingOverlay: newValue })),
}))
