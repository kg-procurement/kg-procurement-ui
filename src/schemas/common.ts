import { z } from 'zod'

// PaginationSpec schema
// TODO: Will probably have to change this to something else when the backend inevitable changes the format
export const paginationSpecSchema = z.object({
  Metadata: z.object({
    TotalPage: z.number(),
    CurrentPage: z.number(),
    TotalEntries: z.number(),
  }),
})

export type PaginationSpec = z.infer<typeof paginationSpecSchema>

export interface PaginationArgs {
  /**
   * One-based indexing for page number to be fetched
   * @default 1
   */
  page?: number
  /**
   * Ascending or descending
   * `ASC` | `DESC`
   * @default `ASC`
   */
  order?: string
  order_by?: string
}
