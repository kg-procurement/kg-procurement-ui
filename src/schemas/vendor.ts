import { z } from 'zod'

export const vendorSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string(),
  bp_id: z.string(),
  bp_name: z.string(),
  rating: z.number(),
  area_group_id: z.string(),
  area_group_name: z.string(),
  sap_code: z.string(),
  modified_date: z.string().datetime(),
  modified_by: z.string(),
  dt: z.string().datetime(),
})

export type Vendor = z.infer<typeof vendorSchema>
