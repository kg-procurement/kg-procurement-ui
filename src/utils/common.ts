export const noop = () => {}

export const formatPrice = (
  amount: number,
  quantityPrice: number,
  currencyCode: string,
  uomName: string
): string => {
  const formattedPrice = new Intl.NumberFormat(
    currencyCode === 'IDR' ? 'id-ID' : 'en-US',
    {
      style: 'currency',
      currency: currencyCode,
    }
  ).format(amount);

  return quantityPrice > 1
    ? `${formattedPrice} / ${quantityPrice} ${uomName}`
    : `${formattedPrice} / ${uomName}`;
};
