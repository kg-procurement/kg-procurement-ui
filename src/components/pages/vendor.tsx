import { useEffect, useState } from 'react'

import { Input } from '@/components/atoms/input.tsx'
import { Typography } from '@/components/atoms/typography.tsx'
import VendorTable from '@/components/features/vendor-table.tsx'
import { Footer } from '@/components/molecules/footer.tsx'
import PageHeader from '@/components/molecules/page-header.tsx'
import { useGetVendorsQuery } from '@/lib/redux/features/vendor/api.ts'
import { useErrorToast } from '@/lib/redux/hooks.ts'
import { useCommonStore } from '@/lib/zustand/common.ts'

export default function VendorPage() {
  const [page, setPage] = useState<number>(1)
  const [productFilter, setProductFilter] = useState('')
  const [locationFilter, setLocationFilter] = useState('')
  const { setShowLoadingOverlay } = useCommonStore()
  const { vendors, metadata, isSuccess, error } = useGetVendorsQuery(
    { page, product: productFilter, location: locationFilter },
    {
      selectFromResult: result => ({
        ...result,
        vendors: result.data?.vendors,
        metadata: result.data?.metadata,
      }),
    },
  )

  useErrorToast(error)

  useEffect(() => {
    setShowLoadingOverlay(!isSuccess)
  }, [isSuccess, setShowLoadingOverlay])

  return (
    <div className="flex min-h-screen w-full flex-col gap-10">
      <PageHeader>
        <Typography variant="h2" className="text-white">
          Search Vendor
        </Typography>
        <Typography variant="body1" className="text-white">
          Lorem Ipsum
        </Typography>
        <div className="flex justify-center gap-4">
          <Input
            placeholder="Filter by Product"
            value={productFilter}
            onChange={e => setProductFilter(e.target.value)}
            className="w-1/2"
          />
          <Input
            placeholder="Filter by Location"
            value={locationFilter}
            onChange={e => setLocationFilter(e.target.value)}
            className="w-1/2"
          />
        </div>
      </PageHeader>
      <div className="flex w-full justify-center">
        {vendors && metadata && isSuccess && (
          <VendorTable
            vendors={vendors}
            metadata={metadata}
            page={page}
            setPage={setPage}
          />
        )}
      </div>
      <Footer />
    </div>
  )
}
