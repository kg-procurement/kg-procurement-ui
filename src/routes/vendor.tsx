import { CheckedState } from '@radix-ui/react-checkbox'
import { createFileRoute, useSearch } from '@tanstack/react-router'
import { useEffect, useState } from 'react'

import Dropdown from '@/components/atoms/dropdown.tsx'
import { Input } from '@/components/atoms/input.tsx'
import { Typography } from '@/components/atoms/typography.tsx'
import VendorTable from '@/components/features/vendor-table.tsx'
import PageHeader from '@/components/molecules/page-header.tsx'
import {
  useGetLocationsQuery,
  useGetVendorsQuery,
} from '@/lib/redux/features/vendor/api.ts'
import { useQueryErrorHandler } from '@/lib/redux/hooks.ts'
import { useCommonStore } from '@/lib/zustand/common.ts'
import { Vendor, vendorPageSearchSchema } from '@/schemas/vendor.ts'

export const Route = createFileRoute('/vendor')({
  component: VendorPage,
  validateSearch: vendorPageSearchSchema,
})

export default function VendorPage() {
  const {
    page: searchPage,
    product_name: searchProduct,
    location: searchLocation,
  } = useSearch({ from: '/vendor' })
  const [page, setPage] = useState<number>(searchPage)
  const [productFilter, setProductFilter] = useState(searchProduct)
  const [locationFilter, setLocationFilter] = useState(searchLocation)
  const { setShowLoadingOverlay } = useCommonStore()
  const [sortBy, setSortBy] = useState<
    'name' | 'area_group_name' | 'rating' | ''
  >('')
  const [sortOrder, setSortOrder] = useState<'asc' | 'desc'>('asc')

  const { vendors, metadata, isSuccess, error } = useGetVendorsQuery(
    {
      page,
      product: productFilter,
      location: locationFilter,
      order_by: sortBy,
      order: sortOrder,
    },
    {
      selectFromResult: result => ({
        ...result,
        vendors: result.data?.vendors,
        metadata: result.data?.metadata,
      }),
    },
  )

  const {
    data: locationsData,
    isSuccess: isLocationsSuccess,
    error: locationsError,
  } = useGetLocationsQuery()

  useQueryErrorHandler(error)
  useQueryErrorHandler(locationsError)

  useEffect(() => {
    setShowLoadingOverlay(!isSuccess && !isLocationsSuccess)
  }, [isSuccess, isLocationsSuccess, setShowLoadingOverlay])

  const locationOptions =
    locationsData?.locations.map(location => ({
      value: location,
      label: location
        .split(' ')
        .map(
          word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase(),
        )
        .join(' '),
    })) || []

  const [vendorIds, setVendorIds] = useState<Set<string>>(new Set())
  const handleUpdateChosenVendor = (
    checked: CheckedState,
    vendorId: string,
  ) => {
    const udpatedSet = new Set(vendorIds)
    if (checked) udpatedSet.add(vendorId)
    else udpatedSet.delete(vendorId)
    setVendorIds(udpatedSet)
  }
  const chooseAllVendor = (vendors: Vendor[], toggle: boolean) => {
    const updatedSet = new Set(vendorIds)
    if (toggle) vendors.forEach(vendor => updatedSet.add(vendor.id))
    else vendors.forEach(vendor => updatedSet.delete(vendor.id))
    setVendorIds(updatedSet)
  }

  const handleSort = (column: 'name' | 'area_group_name' | 'rating') => {
    setSortOrder(sortOrder === 'asc' ? 'desc' : 'asc')
    setSortBy(column)
  }

  return (
    <div className="flex w-full flex-col gap-10">
      <PageHeader>
        <Typography variant="h2" className="text-white">
          Search Vendor
        </Typography>
        <div className="flex w-3/4 justify-between">
          <div className="flex flex-grow justify-stretch">
            <Input
              className="w-full flex-grow"
              placeholder="Filter by Product"
              value={productFilter}
              onChange={e => setProductFilter(e.target.value)}
            />
          </div>
          <div className="flex flex-grow-0 items-center justify-stretch">
            <Dropdown
              options={locationOptions}
              onSelect={(selectedValue) => {
                setLocationFilter(selectedValue)
              }}
              name="Location"
              data-testid="dropdown-button"
            />
          </div>
        </div>
      </PageHeader>
      <div className="flex w-full justify-center">
        {vendors && metadata && isSuccess && (
          <VendorTable
            vendors={vendors}
            metadata={metadata}
            page={page}
            setPage={setPage}
            vendorIds={vendorIds}
            chooseAllVendor={chooseAllVendor}
            handleUpdateChosenVendor={handleUpdateChosenVendor}
            handleSort={handleSort}
            sortBy={sortBy}
            sortOrder={sortOrder}
          />
        )}
      </div>
    </div>
  )
}
