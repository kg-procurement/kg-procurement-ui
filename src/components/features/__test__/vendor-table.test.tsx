import { render, screen } from '@testing-library/react'

import { withWrappers } from '@/lib/testing/utils.tsx'
import { PaginationSpec } from '@/schemas/common.ts'
import { Vendor } from '@/schemas/vendor.ts'

import VendorTable from '../vendor-table.tsx'

const mockVendorsData: Vendor[] = [
  {
    id: '2513',
    name: 'Asaba Computer Centre',
    description: 'Miscrosoft Gold Partner',
    bp_id: '2513',
    bp_name: 'Asaba Computer Centre',
    rating: 1,
    area_group_id: '1',
    area_group_name: 'Indonesia',
    sap_code: '102722',
    modified_date: '2022-12-05T11:47:24Z',
    modified_by: '7',
    dt: '2024-08-28T00:00:00Z',
  },
  {
    id: '2523',
    name: 'Best Denki Indonesia, PT',
    description: '0',
    bp_id: '2523',
    bp_name: 'Best Denki Indonesia, PT',
    rating: 0,
    area_group_id: '1',
    area_group_name: 'Indonesia',
    sap_code: 'None',
    modified_date: '2020-10-29T13:39:12Z',
    modified_by: '0',
    dt: '2024-08-28T00:00:00Z',
  },
  {
    id: '2571',
    name: 'GitHub, Inc',
    description: '0',
    bp_id: '2571',
    bp_name: 'GitHub, Inc',
    rating: 0,
    area_group_id: '1',
    area_group_name: 'Indonesia',
    sap_code: 'None',
    modified_date: '2020-10-29T13:39:12Z',
    modified_by: '0',
    dt: '2024-08-28T00:00:00Z',
  },
  {
    id: '4487',
    name: 'Oktirin Nindy Citrasari / Show Management',
    description: 'None',
    bp_id: '8416',
    bp_name: 'Oktirin Nindy Citrasari / Show Management',
    rating: 0,
    area_group_id: '1',
    area_group_name: 'Indonesia',
    sap_code: 'None',
    modified_date: '2024-09-12T16:57:18Z',
    modified_by: '1075',
    dt: '2024-09-12T00:00:00Z',
  },
]

const mockMetadata: PaginationSpec['Metadata'] = {
  TotalPage: 10,
  CurrentPage: 1,
  TotalEntries: 100,
}

describe('<VendorTable/>', () => {
  it('should render all text provided', () => {
    const { container } = render(
      withWrappers(
        <VendorTable vendors={mockVendorsData} metadata={mockMetadata} />,
      ),
    )
    expect(container.textContent).toMatchSnapshot()
  })
  it('should render table body properly', () => {
    const { container } = render(
      withWrappers(
        <VendorTable vendors={mockVendorsData} metadata={mockMetadata} />,
      ),
    )
    // Check first row data (after header so the index is offset by 1)
    const firstRow = container.querySelectorAll('tr')[1]
    const data = firstRow.querySelectorAll('td')
    expect(data[1].innerText).toMatchInlineSnapshot(`"Asaba Computer Centre"`)
    expect(data[2].innerText).toMatchInlineSnapshot(`"Indonesia"`)
    expect(data[3].innerText).toMatchInlineSnapshot(`"1"`)
  })
  it('should render table header properly', () => {
    render(
      withWrappers(
        <VendorTable vendors={mockVendorsData} metadata={mockMetadata} />,
      ),
    )
    expect(screen.getByText('Vendor Name')).toBeInTheDocument()
    expect(screen.getByText('Location')).toBeInTheDocument()
    expect(screen.getByText('Performance Score')).toBeInTheDocument()
  })
  it('should render email selected button', () => {
    render(
      withWrappers(
        <VendorTable vendors={mockVendorsData} metadata={mockMetadata} />,
      ),
    )
    expect(
      screen.getByRole('button', { name: 'Email Selected Vendor' }),
    ).toBeInTheDocument()
  })
})
