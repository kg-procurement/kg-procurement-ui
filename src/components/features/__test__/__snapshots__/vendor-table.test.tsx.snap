// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<VendorTable/> > should render all text provided 1`] = `"Vendor NameLocationPerformance ScoreAsaba Computer CentreIndonesia1...Best Denki Indonesia, PTIndonesia0...GitHub, IncIndonesia0...Oktirin Nindy Citrasari / Show ManagementIndonesia0...Previous123More pagesNext0 of 99 row(s) selectedEmail Selected Vendor"`;
